src/SDiagonal.jl

```diff
@@ -0,0 +1,117 @@
+# Originally contributed by D. Getz (https://github.com/getzdan), M. Schauer
+# at https://github.com/mschauer/Bridge.jl under MIT License
+
+import Base: getindex,setindex!,==,-,+,*,/,\,transpose,ctranspose,convert, size, abs, real, imag, conj, eye, inv
+import Base.LinAlg: ishermitian, issymmetric, isposdef, factorize, diag, trace, det, logdet, expm, logm, sqrtm
+
+@generated function scalem{T, M, N}(a::SMatrix{M,N, T}, b::SVector{N, T})
+    expr = vec([:(a[$j,$i]*b[$i]) for j=1:M, i=1:N])
+    :(SMatrix{M,N,T}($(expr...)))
+end
+@generated function scalem{T, M, N}(a::SVector{M,T}, b::SMatrix{M, N, T})
+    expr = vec([:(b[$j,$i]*a[$j]) for j=1:M, i=1:N])
+    :(SMatrix{M,N,T}($(expr...)))
+end
+
+struct SDiagonal{N,T} <: StaticMatrix{N, N, T}
+    diag::SVector{N,T}
+    SDiagonal{N,T}(diag::SVector{N,T}) where {N,T} = new(diag)
+end    
+diagtype{N,T}(::Type{SDiagonal{N,T}}) = SVector{N,T}
+diagtype{N}(::Type{SDiagonal{N}}) = SVector{N}
+diagtype(::Type{SDiagonal}) = SVector
+
+# this is to deal with convert.jl
+@inline (::Type{SD})(a::AbstractVector) where {SD <: SDiagonal} = SDiagonal(diagtype(SD)(a)) 
+@inline (::Type{SD})(a::Tuple) where {SD <: SDiagonal} = SDiagonal(diagtype(SD)(a)) 
+@inline (::Type{SDiagonal}){N,T}(a::SVector{N,T}) = SDiagonal{N,T}(a) 
+
+@generated function SDiagonal{N,T}(a::SMatrix{N,N,T})
+    expr = [:(a[$i,$i]) for i=1:N]
+    :(SDiagonal{N,T}($(expr...)))
+end
+
+
+convert{N,T}(::Type{SDiagonal{N,T}}, D::SDiagonal{N,T}) = D
+convert{N,T}(::Type{SDiagonal{N,T}}, D::SDiagonal) = SDiagonal{N,T}(convert(SVector{N,T}, D.diag))
+
+size{N}(D::SDiagonal{N}) = (N,N)
+
+function size{N}(D::SDiagonal{N},d::Int)
+    if d<1
+        throw(ArgumentError("dimension must be ≥ 1, got $d"))
+    end
+    return d<=2 ? N : 1
+end
+
+Base.@propagate_inbounds function getindex{N,T}(D::SDiagonal{N,T}, i::Int, j::Int)  
+    @boundscheck checkbounds(D, i, j)
+    if i == j
+        @inbounds return D.diag[i]
+    else
+        zero(T)
+    end
+end
+
+# avoid linear indexing?
+Base.@propagate_inbounds function getindex{N,T}(D::SDiagonal{N,T}, k::Int) 
+    i, j = ind2sub(size(D), k)
+    D[i,j]
+end
+
+ishermitian{T<:Real}(D::SDiagonal{T}) = true
+ishermitian(D::SDiagonal) = all(D.diag .== real(D.diag))
+issymmetric(D::SDiagonal) = true
+isposdef(D::SDiagonal) = all(D.diag .> 0)
+
+factorize(D::SDiagonal) = D
+
+==(Da::SDiagonal, Db::SDiagonal) = Da.diag == Db.diag
+-(A::SDiagonal) = SDiagonal(-A.diag)
++(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag + Db.diag)
+-(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag - Db.diag)
+-(A::SDiagonal, B::SMatrix) = eye(typeof(B))*A - B
+
+*{T<:Number}(x::T, D::SDiagonal) = SDiagonal(x * D.diag)
+*{T<:Number}(D::SDiagonal, x::T) = SDiagonal(D.diag * x)
+/{T<:Number}(D::SDiagonal, x::T) = SDiagonal(D.diag / x)
+*(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag .* Db.diag)
+*(D::SDiagonal, V::SVector) = D.diag .* V
+*(V::SVector, D::SDiagonal) = D.diag .* V
```

Comment:
or rather: this is already a fallback https://github.com/JuliaLang/julia/blob/master/base/linalg/rowvector.jl#L181 which looks like it is no overhead